[WIP] Powershell Provisioner Exit Code Handling

kp2099 opened the pull request:
This change is to add a wrapper around inline powershell commands that are executed in wrapper to catch the exit code correctly and fail the packer build if there is any error when executing the commands

anurag5sh commented on provisioner/powershell/provisioner.go:
```diff
@@ -38,6 +37,33 @@ var psEscape = strings.NewReplacer(
+	Set-PsDebug -Trace {{.DebugMode}}
+	{{- end}}
+	$exitCode = 0
+	{{.Vars}}
```
Do u think we can move `{{.Vars}}` inside try block as well? May help in catching issues if any.

kp2099 replied:
I didn't include {{.Vars}} inside the try-catch block because, if you look at how Vars is injected, it's done through the createFlattenedEnvVars() method. This method already processes (catches issues during rendering) and formats the environment variables appropriately for the target OS and injects them into the payload as a flattened string

That said, adding it inside the try block wouldn’t cause any issues, so we can certainly consider doing that.

anurag5sh replied:
Okay makes sense. But I think we can go ahead and add it inside, it acts like a fail check for any changes made in the `createFlattenedEnvVars()` in future.

kp2099 commented:
> Overall the changes look good to me. But we can test this in few more scenarios or probably add test cases. Some of which I can think of:
> 
> 1. Any form of syntax errors in the powershell script
> 2. Certain powershell commands like `Get-Item "C:\nonexistent.txt"` . External programs will error out, but we can test if the powershell cmdlet also behaves the same
> 3. A nested try/catch scenario

To clarify, PowerShell errors are generally categorized into two types: terminating and non-terminating errors. The wrapper introduced here is designed to capture terminating errors and return their corresponding exit codes.

A terminating error includes:

1. Any invalid syntax.
2.  CommandNotFoundException (occurs when attempting to run a non-existent command).
3. Runtime exceptions (e.g., division by zero).
4.  Explicitly thrown exceptions.

Regarding the second test case, `Get-Item "C:\nonexistent.txt"` produces a non-terminating error by default. This means the script would continue execution without failing and return an exit status of zero. However, if users want to explicitly handle non-terminating errors and return a non-zero exit code, they can do so using the ErrorActionPreference variable. Added it in the acceptance test for reference.

There was some discussion (https://github.com/hashicorp/packer/issues/6449) about whether Packer should enforce ErrorActionPreference='Stop' by default. However, since this behavior differs from PowerShell's default setting, we have opted to leave it up to the user to configure based on their needs.

Additionally, I have incorporated test cases for the nested try/catch scenario.

kp2099 commented:
> LGTM Also please test the case where ExecutionPolicy is set to None.

Added the test case

JenGoldstrich commented on provisioner/powershell/provisioner_test.go:
```diff
@@ -35,13 +36,15 @@ func TestProvisionerPrepare_extractScript(t *testing.T) {
 	// File contents should contain 2 lines concatenated by newlines: foo\nbar
 	readFile, err := os.ReadFile(file)
-	expectedContents := "foo\nbar\n"
+	expectedContents := "if (Test-Path variable:global:ProgressPreference) {\n         set-variable -name variable:global:ProgressPreference -value 'SilentlyContinue'\n         }\n         \n         $exitCode = 0\n         try {\n         $env:PACKER_BUILDER_TYPE=\"\"; $env:PACKER_BUILD_NAME=\"\"; \n         foobar\n         $exitCode = 0\n         } catch {\n         Write-Error \"An error occurre ...
```
This is just a nit pick, but I would break this up into several lines of text joined together, just to make this expected contents a bit more readable

JenGoldstrich reviewed (approved):
This looks good to me, I've tested it with some basic templates, thank you for your work on this one @kp2099!

Let's plan to release this in the next minor release since it is a slight breaking change

mloskot commented:
@kp2099  Thank very much you for this fix!
